### PR TITLE
Adding EssentialsWarpEvent

### PR DESCRIPTION
**Pull request by Banbeucmas**
Tho you can still use PlayerTeleportEvent for this, I think it is better to have an event especially for EssentialsWarp for checking warps tho.
This should help a lot on dependency which are trying to hook into the plugin.

### CODE REVIEW

**Review by Dolphin2Point1 — APPROVED** [review]
Great! Some Plugins could really use this for an api!
Edit: I have compiled it for use. If you're too lazy, look here: https://github.com/loper12/BanbeucmasEssPatch

File: Essentials/src/com/earth2me/essentials/Teleport.java [hunk 1]
```diff
@@ -247,6 +249,11 @@ protected void respawnNow(IUser teleportee, TeleportCause cause) throws Exceptio
     //The warp function is a wrapper used to teleportPlayer a player to a /warp
     @Override
     public void warp(IUser teleportee, String warp, Trade chargeFor, TeleportCause cause) throws Exception {
+        EssentialsWarpEvent event = new EssentialsWarpEvent(teleportee, warp);
+        Bukkit.getServer().getPluginManager().callEvent(event);
+        if(event.isCancelled()){
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/Teleport.java]: This should be spaced out.

**Review by mdcfe — COMMENTED** [review]
This mostly looks good, but I haven't been able to test it so I can't approve this yet.

File: Essentials/src/com/earth2me/essentials/Teleport.java (hunk 1 above)
Comment (Banbeucmas) [reply Essentials/src/com/earth2me/essentials/Teleport.java]: Testing with Remote Debugging and also Checkpoints, the event was called. I haven't test the listeners tho

File: Essentials/src/net/ess3/api/events/EssentialsWarpEvent.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+
+public class EssentialsWarpEvent extends Event implements Cancellable {
+    private static final HandlerList handlers = new HandlerList();
+
+    private IUser user;
+    private String warp;
+    private boolean cancelled = false;
+
```
Comment (Banbeucmas) [inline Essentials/src/net/ess3/api/events/EssentialsWarpEvent.java]: Should I add Trade also?

Comment (mdcfe) [reply Essentials/src/net/ess3/api/events/EssentialsWarpEvent.java]: Could do, I suppose.

Comment (Banbeucmas) [reply Essentials/src/net/ess3/api/events/EssentialsWarpEvent.java]: Ok wait for me a moment

**Comment by mdcfe** [comment]
We could also add a method `EssentialsWarpEvent#setWarp` which changes the warp destination, then retrieve the warp destination from the event in `Teleport#warp`.

Also, prefixing the event with `Essentials` isn't consistent with other Essentials events - I'd suggest renaming it to `UserWarpEvent`, but open to any other suggestions.

**Comment by Banbeucmas** [comment]
> We could also add a method EssentialsWarpEvent#setWarp which changes the warp destination, then retrieve the warp destination from the event in Teleport#warp.

Like that?
I tested and pretty sure it worked flawlessly without errors

File: Essentials/src/net/ess3/api/events/UserWarpEvent.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    public String getWarp() {
+        return warp;
+    }
+
+    public Trade getTrade() {
+        return trade;
+    }
+
+    public void setWarp(String warp){
```
Comment (mdcfe) [inline Essentials/src/net/ess3/api/events/UserWarpEvent.java]: Spacing here too: `public void setWarp(String warp) {`

File: Essentials/src/com/earth2me/essentials/Teleport.java [hunk 4]
```diff
@@ -247,6 +249,13 @@ protected void respawnNow(IUser teleportee, TeleportCause cause) throws Exceptio
     //The warp function is a wrapper used to teleportPlayer a player to a /warp
     @Override
     public void warp(IUser teleportee, String warp, Trade chargeFor, TeleportCause cause) throws Exception {
+        UserWarpEvent event = new UserWarpEvent(teleportee, warp, chargeFor);
+        Bukkit.getServer().getPluginManager().callEvent(event);
+
+        if(event.isCancelled()){
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/Teleport.java]: This should still be spaced out: `if (event.isCancelled()) {`

Comment (Banbeucmas) [reply Essentials/src/com/earth2me/essentials/Teleport.java]: Ok I will do it tonight

File: Essentials/src/net/ess3/api/events/UserWarpEvent.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+    public IUser getUser() {
+        return user;
+    }
+
+    public String getWarp() {
+        return warp;
+    }
+
+    public Trade getTrade() {
```
Comment (drtshock) [inline Essentials/src/net/ess3/api/events/UserWarpEvent.java:39]: Perhaps some javadocs would help?

Comment (Banbeucmas) [reply Essentials/src/net/ess3/api/events/UserWarpEvent.java:39]: Should help. I am not that familiar with Javadocs tho

**Comment by Banbeucmas** [comment]
@md678685 @drtshock Done, can you guys check for what is left? 
If it is ok then I think we can merge it by now.

File: Essentials/src/net/ess3/api/events/UserWarpEvent.java (hunk 5 above)
Comment (Banbeucmas) [reply Essentials/src/net/ess3/api/events/UserWarpEvent.java:39]: I don't know if I were to left documentation there, providing that all of other Events doesn't have any documentation except the heading of the class.

**Review by mdcfe — APPROVED** [review]

**Comment by mdcfe** [comment]
This has been merged into the 1.13 branch and is now part of experimental builds.

**Comment by Banbeucmas** [comment]
Thanks for the merge.
Guess I will back to my work then.

**Pull request merged**